pkg/internal/request/span.go

```diff
@@ -63,36 +112,184 @@ type PidInfo struct {
 // REMINDER: any attribute here must be also added to the functions SpanOTELGetters,
 // SpanPromGetters and getDefinitions in pkg/internal/export/metric/definitions.go
 type Span struct {
-	Type           EventType
-	IgnoreSpan     IgnoreMode
-	Method         string
-	Path           string
-	Route          string
-	Peer           string
-	PeerPort       int
-	Host           string
-	HostPort       int
-	Status         int
-	ContentLength  int64
-	RequestStart   int64
-	Start          int64
-	End            int64
-	ServiceID      svc.ID // TODO: rename to Service or ResourceAttrs
-	TraceID        trace2.TraceID
-	SpanID         trace2.SpanID
-	ParentSpanID   trace2.SpanID
-	Flags          uint8
-	Pid            PidInfo
-	PeerName       string
-	HostName       string
-	OtherNamespace string
-	Statement      string
+	Type           EventType      `json:"type"`
+	IgnoreSpan     IgnoreMode     `json:"ignoreSpan"`
+	Method         string         `json:"-"`
+	Path           string         `json:"-"`
+	Route          string         `json:"-"`
+	Peer           string         `json:"peer"`
+	PeerPort       int            `json:"peerPort,string"`
+	Host           string         `json:"host"`
+	HostPort       int            `json:"hostPort,string"`
+	Status         int            `json:"status,string"`
+	ContentLength  int64          `json:"-"`
+	RequestStart   int64          `json:"-"`
+	Start          int64          `json:"-"`
+	End            int64          `json:"-"`
+	ServiceID      svc.ID         `json:"-"` // TODO: rename to Service or ResourceAttrs
+	TraceID        trace2.TraceID `json:"traceID"`
+	SpanID         trace2.SpanID  `json:"spanID"`
+	ParentSpanID   trace2.SpanID  `json:"parentSpanID"`
+	Flags          uint8          `json:"flags,string"`
+	Pid            PidInfo        `json:"-"`
+	PeerName       string         `json:"peerName"`
+	HostName       string         `json:"hostName"`
+	OtherNamespace string         `json:"-"`
+	Statement      string         `json:"-"`
 }
 
 func (s *Span) Inside(parent *Span) bool {
 	return s.RequestStart >= parent.RequestStart && s.End <= parent.End
 }
 
+const (
+	kClient   = "CLIENT"
+	kServer   = "SERVER"
+	kProducer = "PRODUCER"
+	kConsumer = "CONSUMER"
+	kInternal = "INTERNAL"
+)
+
+func kindString(span *Span) string {
+	switch span.Type {
+	case EventTypeHTTP, EventTypeGRPC, EventTypeKafkaServer, EventTypeRedisServer:
+		return kServer
+	case EventTypeHTTPClient, EventTypeGRPCClient, EventTypeSQLClient, EventTypeRedisClient:
+		return kClient
+	case EventTypeKafkaClient:
+		switch span.Method {
+		case MessagingPublish:
+			return kProducer
+		case MessagingProcess:
+			return kConsumer
+		}
+	}
+	return kInternal
+}
+
+// helper attribute functions used by JSON serialization
+type SpanAttributes map[string]string
+
+// nolint
```

Comment:
What was the reason you disabled the linter here? When the linter complains about code it means it's not "go like" enough. For Go, a lot of the style is actually coding convention, so blanket disabling of the linter is not a good idea. Sometimes it's ok to disable the linter for example for cyclomatic complexity, if breaking down the function will make things suboptimal.

Reply:
It wanted to me to turn single case switches into an if statement, which makes sense overall (Kernighan talks about it extensively in the Practice of Programming), but not here - as you can see there's some sort of pattern at play, and replace those switches with if statements would decrease clarity in this particular case.
It also complained about cyclomatic complexity - but in this case I make the conscious decision of wrapping the multiple *Attribute(...) functions from the former revisions of this PR into a single self-contained function, again, for clarity and to reduce duplication.
I assume there must be a way for me to disable just these two particular checks. I will RTFM and make the changes.

Reply:
I understand now, but to be honest I don't see how switches make this code any more clear. Essentially, if I tried to understand what attributes are applied to EventTypeHTTP as a reader of this code, I'd have hard time with the switches and with the ifs. At the same time, the generated code will be large number of conditional branches.

A better way to do this would be by using the same approach as traceAttributes https://github.com/grafana/beyla/blob/805db0338da6af625c6a42f1bb6d519e8e05153a/pkg/export/otel/traces.go#L531. It also makes me wonder if we cannot somehow reuse that code?

Reply:
That's much better! Thanks, I will make the change!